File: src/main/java/de/achimonline/kickassembler/acbg/settings/KickAssemblerSettingsConfigurable.java
package de.achimonline.kickassembler.acbg.settings;

import com.intellij.openapi.options.Configurable;
import com.intellij.openapi.options.ConfigurationException;
import de.achimonline.kickassembler.acbg.properties.KickAssemblerProperties;
import org.jetbrains.annotations.Nls;
import org.jetbrains.annotations.Nullable;

import javax.swing.*;

public class KickAssemblerSettingsConfigurable implements Configurable {
    private KickAssemblerSettingsPane kickAssemblerSettingsPane;

    @Nls(capitalization = Nls.Capitalization.Title)
    @Override
    public String getDisplayName() {
        return KickAssemblerProperties.message("settings.name");
    }

    @Nullable
    @Override
    public JComponent createComponent() {
        if (kickAssemblerSettingsPane == null) {
            kickAssemblerSettingsPane = new KickAssemblerSettingsPane();
        }

        return kickAssemblerSettingsPane.getPanel();
    }

    @Override
    public boolean isModified() {
        return kickAssemblerSettingsPane != null && kickAssemblerSettingsPane.isModified();
    }

    @Override
    public void apply() throws ConfigurationException {
        kickAssemblerSettingsPane.storeSettings(KickAssemblerSettingsService.getInstance().getState());
    }

    @Override
    public void reset() {
        kickAssemblerSettingsPane.setData(KickAssemblerSettingsService.getInstance().getState());
    }

    @Override
    public void disposeUIResources() {
    }
}

File: src/main/kotlin/de/achimonline/kickassembler/acbg/settings/KickAssemblerSettingsConfigurable.kt
package de.achimonline.kickassembler.acbg.settings

import com.intellij.execution.ui.DefaultJreSelector
import com.intellij.execution.ui.JrePathEditor
import com.intellij.openapi.options.BoundConfigurable
import com.intellij.openapi.ui.DialogPanel
import com.intellij.openapi.util.Pair
import com.intellij.ui.dsl.builder.AlignX
import com.intellij.ui.dsl.builder.panel
import com.intellij.ui.dsl.builder.toMutableProperty
import de.achimonline.kickassembler.acbg.properties.KickAssemblerProperties.message

class KickAssemblerSettingsConfigurable : BoundConfigurable(message("settings.name")) {
    private val jrePathEditor = JrePathEditor(object : DefaultJreSelector() {
        override fun getNameAndDescription(): Pair<String?, String> {
            return Pair.empty()
        }
    })

    private val settings
        get() = KickAssemblerSettingsState.instance.settings

    override fun createPanel(): DialogPanel {
        return panel {
            group(message("settings.jre")) {
                row {
                    cell(jrePathEditor)
                        .align(AlignX.FILL)
                        .bind(
                            JrePathEditor::getJreNameOrPath,
                            JrePathEditor::setJreNameOrPath,
                            settings::jreNameOrPath.toMutableProperty()
                        )
                }.rowComment(message("settings.jre.comment"))
            }
        }
    }
}
